core-rust/core-api-server/src/core_api/conversions/substates/access_controller.rs

```diff
@@ -12,11 +13,95 @@ pub fn to_api_access_controller_substate(
+            controlled_vault_entity: Box::new(to_api_entity_reference(
+                context,
+                controlled_asset.as_node_id()
+            )?),
+            timed_recovery_delay_minutes: timed_recovery_delay_in_minutes
+                .as_ref()
+                .map(|minutes| to_api_u32_as_i64(*minutes)),
+            recovery_badge_resource_address: to_api_resource_address(context, recovery_badge)?,
+            is_primary_role_locked: matches!(
+                primary_role_locking_state,
+                PrimaryRoleLockingState::Locked
+            ),
+            primary_role_recovery_attempt: match primary_role_recovery_attempt_state {
+                PrimaryRoleRecoveryAttemptState::NoRecoveryAttempt => None,
+                PrimaryRoleRecoveryAttemptState::RecoveryAttempt(recovery_proposal) =>
+                    Some(Box::new(models::PrimaryRoleRecoveryAttempt {
+                        recovery_proposal: Box::new(to_api_recovery_proposal(
+                            context,
+                            recovery_proposal
+                        )?)
+                    })),
+            },
+            has_primary_role_badge_withdraw_attempt: matches!(
+                primary_role_badge_withdraw_attempt_state,
+                PrimaryRoleBadgeWithdrawAttemptState::BadgeWithdrawAttempt
+            ),
+            recovery_role_recovery_attempt: match recovery_role_recovery_attempt_state {
+                RecoveryRoleRecoveryAttemptState::NoRecoveryAttempt => None,
+                RecoveryRoleRecoveryAttemptState::RecoveryAttempt(attempt_state) => {
+                    let (recovery_proposal, timed_recovery_allowed_after) = match attempt_state {
+                        RecoveryRoleRecoveryState::UntimedRecovery(proposal) => (proposal, None),
+                        RecoveryRoleRecoveryState::TimedRecovery {
+                            proposal,
+                            timed_recovery_allowed_after,
+                        } => (proposal, Some(timed_recovery_allowed_after)),
+                    };
+                    Some(Box::new(models::RecoveryRoleRecoveryAttempt {
+                        recovery_proposal: Box::new(to_api_recovery_proposal(
+                            context,
+                            recovery_proposal,
+                        )?),
+                        timed_recovery_allowed_after: timed_recovery_allowed_after
+                            .map(to_api_instant)
+                            .transpose()?
+                            .map(Box::new),
+                    }))
+                }
+            },
+            has_recovery_role_badge_withdraw_attempt: matches!(
+                recovery_role_badge_withdraw_attempt_state,
+                RecoveryRoleBadgeWithdrawAttemptState::BadgeWithdrawAttempt
+            ),
         }
     ))
 }
+
+pub fn to_api_recovery_proposal(
+    context: &MappingContext,
+    substate: &RecoveryProposal,
```

Comment:
Very minor - but this shouldn't be called substate.

And we generally should be using an explicit destructuring so that if new fields are added, we have a compile error and are prompted to add them.